Transform struct

CodeDoes opened the pull request:
Please look at this and tell me what you think. I'd like to improve.

CodeDoes committed:
Adds up all the mouse wheel events.
fixed some spacing.

CodeDoes committed:
fixed errors

CodeDoes committed:
Added a calcDim proc for Entity

CodeDoes committed:
Merge branch 'devel' into devel

CodeDoes committed:
limited rotation ammount to max 360.0

CodeDoes committed:
Applied requested changes

CodeDoes committed:
Merge branch 'devel' of github.com:CodeDoes/nimgame2 into devel

CodeDoes committed:
Added transform and all related procs created so far

CodeDoes committed:
Merge branch 'devel' of https://github.com/Vladar4/nimgame2 into devel

# Conflicts:
#	nimgame2/entity.nim

CodeDoes committed:
moved the procs into more appropriate modules

CodeDoes committed:
Fixed the looping import error :|

Vladar4 commented on nimgame2/entity.nim:
```diff
@@ -821,5 +821,71 @@ method onCollide*(entity, target: Entity) {.base.} =
   discard
 
 
+proc transform*(self:Entity):Transform=
```
General formatting style notes, here and further:
* change `self` to `entity`, etc. to conform to the established style;
* procs and templates' names should be in camelCase;
* don't forget spaces after `:` and before `=`;
* two-line spacing between procs, except the first proc after the section comment (one line then);
* consider using templates instead of procs where it seems reasonable (for example, one-line procedures, etc.).

So, e.g. this line should be changed into `template transform*(entity: Entity): Transform =`, etc.

CodeDoes replied:
I'll format the PR

Vladar4 commented on nimgame2/entity.nim:
```diff
@@ -821,5 +821,71 @@ method onCollide*(entity, target: Entity) {.base.} =
+    angle: self.absRot, 
+    scale: self.absScale).Transform
+
+proc apply_relative_transform*(self: var Entity, transform: Transform)=
```
Rename into `` `transform=`*``.

Vladar4 commented on nimgame2/entity.nim:
```diff
@@ -821,5 +821,71 @@ method onCollide*(entity, target: Entity) {.base.} =
+proc rect*(self: Entity):Rect=
+  return self.graphic.rect(self.center * self.scale)
+
+proc relative_center*(self:Entity): Coord= 
```
Considering similarities between `relative_` and `inverse_` variants, how about merging them together with `inverse` as a boolean argument that defaults to `false`?

CodeDoes replied:
I don't feel comfortable with adding ambiguity. center(true) is not nearly as descriptive as relative_center or inverse_center. 
For me inverse means "what is the world point" relative means "without rotation or scale or displacement what would that point be"

CodeDoes replied:
Maybe only keep the "relative" points. remove the world "relative" from the points. then if someone wants the world location of a relative point they can just go `ent.transform.point(ent.topleft)` or something

Vladar4 replied:
> Maybe only keep the "relative" points. remove the world "relative" from the points. then if someone wants the world location of a relative point they can just go ent.transform.point(ent.topleft) or something

I like this idea even more.

Vladar4 commented on nimgame2/types.nim:
```diff
@@ -575,3 +575,15 @@ converter toCoord*(d: Dim): Coord =
+#  Transform  #
+#=============#
+
+proc apply*(self: var Transform, source: Transform)=
```
Rename into `` `=`* ``.

CodeDoes replied:
fixing

CodeDoes replied:
To do this Transform needs to be an object.

Vladar4 replied:
Oh, sorry. For the tuples this proc isn't needed at all. You could just
```
var src, dst: Ttransform
...
dst = src
```

Vladar4 commented on nimgame2/utils.nim:
```diff
@@ -551,3 +551,27 @@ proc update*(counter: Counter) =
 proc value*(counter: Counter): int {.inline.} =
   counter.current
 
+#===========#
```
I feel that this section belongs into `types` module.

What do you think about merging last three procs into one, e.g.:

```
proc `+=`*(transform: Transform, pos: Coord = (0.0, 0.0), angle: Angle = 0.0, scale: Scale = 0.0): Transform =
  ...
```
?

CodeDoes replied:
Scaling isn't `+` its `*`
The whole idea is to have a temporary transform. Something like. "What would be my transform if I was moved 5 steps forward and turned 12 degrees: `transform.translated((5.0,0.0)).rotated(12.0)`

CodeDoes replied:
The reason its in utils is because of `rotate`

Vladar4 replied:
With such procedure chaining you are copying the transform tuple each time though, it might hit the performance…

May be something akin to
```
proc transformed*(target: Transform, translated: Coord = (0.0, 0.0), rotated: Angle = 0.0, scaled: Scale = 1.0): Transform =
  ...

target.transformed(translated = (5.0, 0.0), rotated = (12.0))
```
then?

Vladar4 commented:
OK, so above are my comments regarding this pull request. Let me know what you think.

BTW, do you have some practical usage example of this?

Thanks!

CodeDoes commented:
Yes I was using it to make rotated text. For some reason my code is really slow. I know polygon collision checking isn't efficient but it shouldn't be as slow as I'm seeing.
I also used it to rotate polygons. I'm not sure if that's considered practical. My goal is to eventually build a small animation program

CodeDoes committed:
moved Transform procs to types as suggested.
changed the mass of procs to templates. (as many as I could)
changed Transform.point(Coord) to Transform * Coord
removed translated,scaled,rotated

CodeDoes committed:
Small demo using transform a bit

CodeDoes committed:
updated example. Fixed lag?

CodeDoes committed:
moved modules to private folder

Vladar4 commented:
> Yes I was using it to make rotated text. For some reason my code is really slow. I know polygon collision checking isn't efficient but it shouldn't be as slow as I'm seeing.

Can you link me to a basic usage example then?

CodeDoes committed:
Changed Transform attribute called inverse to local .

CodeDoes commented:
I pushed an example. ... Didn't realise I was committing but not pushing back to the remote

CodeDoes commented:
@Vladar4

CodeDoes committed:
removed apply and copy proc

CodeDoes committed:
Fixed formatting and parameter naming

Vladar4 commented on demos/README.md:
```diff
@@ -45,3 +45,4 @@ Nimgame 2 Demos
 
 * [demo22](demo22) MpegGraphic
 
+* [demo23](demo23) PolygonGraphic
```
Considering there are no actual `PolygonGraphic` type in the code, could you change the title (maybe for "Transform", etc.)?

Vladar4 commented on demos/demo23/demo23.code-workspace:
```diff
@@ -0,0 +1,15 @@
+{
```
This file shouldn't be here.

Vladar4 commented on demos/demo23/private/circle_graphic.nim:
```diff
@@ -0,0 +1,62 @@
+method dim*(self:CircleGraphic):Dim=
+  return (int self.radius*2,int self.radius*2)
+
+
```
A lot of unneeded empty lines here.

Vladar4 commented on demos/demo23/demo23.nim:
```diff
@@ -0,0 +1,13 @@
+
+game = newGame()
+settings.updateInterval = 1000
+if game.init(400, 400, "TextEntity module test"):
```
Could you change the resolution to 640x480 like all other demos? Plus, engine's info-panel obscures your objects in the top-left corner right now.

Vladar4 reviewed (changes requested):
Thanks, looks quite good for a start. Here's a couple of final notes before I merge this.

CodeDoes committed:
removed vscode file, changed resolution, changed demo name

CodeDoes committed:
changed widget positions

Vladar4 commented:
Awesome!

CodeDoes commented:
Cool. tell me if there's something else I can work on.

On 18 June 2018 at 08:50, Vladimir Arabadzhi <notifications@github.com>
wrote:

> Merged #22 <https://github.com/Vladar4/nimgame2/pull/22>.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Vladar4/nimgame2/pull/22#event-1685443899>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AT59_fmv9K8UrFvslp6oeCoBEydHt5iIks5t903TgaJpZM4UqdK7>
> .
>
